rustual-boy-core/src/v810.rs

```diff
@@ -548,30 +628,38 @@ impl V810 {
                 }),
                 OPCODE_BITS_LDSR => format_ii!(|imm5, reg2| {
                     let value = self.reg_gpr(reg2);
+
+
                     match imm5 {
                         OPCODE_SYSTEM_REGISTER_ID_EIPC => {
+                            trace_add!(TraceEntry::SysReg(imm5, self.reg_eipc));
                             self.reg_eipc = value;
                         }
                         OPCODE_SYSTEM_REGISTER_ID_EIPSW => {
+                            trace_add!(TraceEntry::SysReg(imm5, self.reg_eipsw));
                             self.reg_eipsw = value;
                         }
                         OPCODE_SYSTEM_REGISTER_ID_FEPC => {
+                            trace_add!(TraceEntry::SysReg(imm5, 0));
                             logln!(Log::Cpu, "WARNING: ldsr fepc not yet implemented (value: 0x{:08x})", value);
                         }
                         OPCODE_SYSTEM_REGISTER_ID_FEPSW => {
+                            trace_add!(TraceEntry::SysReg(imm5, 0));
                             logln!(Log::Cpu, "WARNING: ldsr fepsw not yet implemented (value: 0x{:08x})", value);
                         }
                         OPCODE_SYSTEM_REGISTER_ID_ECR => {
+                            trace_add!(TraceEntry::SysReg(imm5, self.reg_ecr as u32));
                             self.reg_ecr = value as _;
                         }
-                        OPCODE_SYSTEM_REGISTER_ID_PSW => self.set_reg_psw(value),
+                        OPCODE_SYSTEM_REGISTER_ID_PSW => {
+                            trace_add!(TraceEntry::SysReg(imm5, self.reg_psw()));
+                            self.set_reg_psw(value);
+                        }
                         OPCODE_SYSTEM_REGISTER_ID_CHCW => {
+                            trace_add!(TraceEntry::SysReg(imm5, self.reg_chcw));
```

Comment:
Sorry if I wasn't very clear here; when I mentioned a function that could get this register's value, I meant for that to still calculate the value on-the-fly as we had previously, instead of operate on a backing variable. Now it looks like there's a function to get its value which is not currently used?

The reason why we use the on-the-fly pattern is it reduces code/logic. For example, with `reg_chcw`, we now have to set its value, _then_ set the cache to be enabled. If we introduced some other code path to change this value, it's quite possible to accidentally only update the `reg_chcw` variable without updating the cache's enabled state, or vice versa; they could get out of sync. This isn't too bad with one register in isolation, but think of how many registers/states we already have in this emulator, and we're not even done! If we only have one code path to set/get these values and the state only lives in one place (in this case, only in the cache, rather than both in the cache and the cpu's variable) we've eliminated a potential source for error and simplified the system as a whole. So I understand why it may be a bit annoying that I'm adamant about that pattern, but I'd really prefer if we stuck to it. I don't see how it shouldn't also work in this case, if you just call the `reg_chcw` fn here instead of referencing a variable, and set it similarly. The `reg_psw` fn's are a good example of this pattern.